pagefaultgames/pokerogue
# [QoL] IV scanner update to hide the prompt and let you see the stats mid battle on the stats screen

## What are the changes?
The IV scanner has been updated in two ways - the first is a setting that will allow you to disable the IV scanner prompt and all associated parts (animation, message etc), and the second is to change the stat container labels to show green/gold for the IV mid battle to show if the stats of the pokemon are better than what you've currently got.

## Why am I doing these changes?
A lot of people on discord have mentioned this is a feature they wanted, so I decided to make it

## What did change?
The first option is a setting that enables or disables the prompt about scanning a pokemon's IV. This is under the "general" tab of the settings.

The second option is to change the labels of the stat container to be green or gold depending on the stat. The stats container will show you the same amount of IV updates that you would from the IV scanner - so if you have no scanners you won't see any updates, all the way up to seeing all 6 stats for 3 scanners. If the stat is better than what you have, it'll show green, whereas if it's perfect IV it'll show gold; this is so if you do have the prompts being skipped, you can still see if the stat is perfect. Lastly, if the stat is unknown (not enough scanners), not scanned (ACC/EVA), or equal/worse than what you have, the label is white as usual.

The last change is to do with the stat labels themselves. Now if you have at least one IV scanner, the enemy stats will show a label for HP. There is no +/- number associated with it, as it's just there purely for IV purposes. This label does not show for player pokemon, and only shows when an IV scan would happen - so this label doesn't show for trainer battles

### Screenshots/Videos
Here is a screenshot of the settings screen showing the new setting:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/fd1a9d34-0b39-4c5b-8be3-129595bb1d51)

Here is a screenshot of the stats screen in battle showing some stats and the IV scanner on screen:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/e6fe13b1-bd68-4a95-a3ae-b6cd918a0316)

As you can see, the only stats that are better than what we have are HP, Def and Sp. Def. Since Def is "Best" (i.e. 31), it shows up gold on the stats container. However, HP and Sp. Def are not "Best", so it shows up green. The Atk for this pokemon isn't better than what we have, so it's white like usual. The other stats are unknown/unscanned, so they also show up as white. This is with 2 IV scanners, so only 4 stats are shown.

Below is an example of a fight where pokemon have changed their stats with various stat changing moves: 

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/73ab32a6-27d4-41b2-9128-508e94465473)

Below is an example with 2 IV scanners in a run showing the difference with the IV scanner setting on (current behaviour) and the IV scanner setting off:

https://github.com/pagefaultgames/pokerogue/assets/66582645/47438007-3edb-4a4e-a0f3-74605671ad12

Here's an example of a run where there's no IV scanners. You'll notice that the HP stat isn't shown, but everything has been moved over a bit. This is so that when the HP stat does get shown, the ui elements don't shift because of it:

https://github.com/pagefaultgames/pokerogue/assets/66582645/4e590d8e-56d3-4c78-bf69-a7d767bd1499

Here's an example with 3 IV scanners:

https://github.com/pagefaultgames/pokerogue/assets/66582645/77a66662-862d-4d34-8aec-2dfc2387d274

Here's a double battle with 1 IV scanner while also going through different waves:

https://github.com/pagefaultgames/pokerogue/assets/66582645/7696524a-6c99-45ee-855f-1be7cec2a94f

Below is an example of fighting a trainer and not seeing the IV scanner prompt:

https://github.com/pagefaultgames/pokerogue/assets/66582645/c2900bdf-7013-4303-87dc-b7e8a188b3cc

Lastly, below is an example where you start with no IV scanners and then pick up on in the store:

https://github.com/pagefaultgames/pokerogue/assets/66582645/4dab40d5-9ed9-47c5-a1f2-ab89ed8c2de8

Finally, here's a screenshot of the IV scanner in action for the legacy UI:

![image](https://github.com/pagefaultgames/pokerogue/assets/66582645/340271b7-6a95-485f-939c-ff5277c6e307)

## How to test the changes?
Download the branch to play with the changes. When you're at the battle screen, you can open the stats container (default Shift/C) to see the changes. Also in the settings, there should be an option to turn the setting on/off to hide the IV prompts

You can also use any of the following overrides:

Line 32 - this will force a double battle:

`export const DOUBLE_BATTLE_OVERRIDE: boolean = true;`

Line 90 - this will let you choose a specific pokemon to fight against. For example, this is set to spawn a charizard, but you can change that to anything else to fight other things:

`export const OPP_SPECIES_OVERRIDE: Species | integer = Species.CHARIZARD;`

Line 96 - this will let you set the moves the enemy opponent has; for example this could be used to set stat changing moves for enemy pokemon to see the stat numbers appear. In this example, we have given them the moves no retreat, quiver dance, and shell smash twice. This can be set to anything:

`export const OPP_MOVESET_OVERRIDE: Array<Moves> = [Moves.NO_RETREAT, Moves.QUIVER_DANCE, Moves.SHELL_SMASH, Moves.SHELL_SMASH];`

Line 99 - this will let you set the IVs of the enemy pokemon. It takes an array of 6 numbers, which, in order, are HP, Atk, Def, Sp. Atk, Sp. Def, Spd:

`export const OPP_IVS_OVERRIDE: integer | integer[] = [31, 24, 5, 3, 19, 16];`

Line 121 - this will give you 3 IV scanners. To change it, change the "count: 3" to "count: 1" or "count: 2":

`export const STARTING_MODIFIER_OVERRIDE: Array<ModifierOverride> = [{name: "IV_SCANNER", count: 3}];`

**NOTE:** If you want to get rid of your IV scanners, you have to revert this; changing the count to 0 will NOT work, so to have 0 IV scanners, you need to use:

`export const STARTING_MODIFIER_OVERRIDE: Array<ModifierOverride> = [];`

Also note that for the IV scanners, you'll need to start a new game to see the effects. At first it'll look like it didn't work, but you'll be given the choice to see the pokemon's IVs before the fight as usual.

## Checklist
